### PR TITLE
License finding curations

### PR DESCRIPTION
**Pull request by fviernau**
has to be merged after https://github.com/heremaps/oss-review-toolkit/pull/1844.

### CODE REVIEW

File: model/src/main/kotlin/config/LicenseFindingCuration.kt [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+/**
+ * A curation for license finding(s) applying to all license findings for which all matchers represented by this classes
+ * properties match. When applied it assigns the [concludedLicense] to all matching findings.
+ */
+data class LicenseFindingCuration(
+    /**
+     * A glob to match the file path of a license finding.
+     */
+    val path: String,
+    /**
```
Comment (mnonnenmacher) [inline model/src/main/kotlin/config/LicenseFindingCuration.kt]: Add an empty line before the comment, same several times below.

File: model/src/main/kotlin/config/LicenseFindingCuration.kt [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+package com.here.ort.model.config
+
+import com.fasterxml.jackson.annotation.JsonInclude
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize
+import com.fasterxml.jackson.databind.annotation.JsonSerialize
+import com.fasterxml.jackson.databind.util.StdConverter
+
+/**
+ * A curation for license finding(s) applying to all license findings for which all matchers represented by this classes
+ * properties match. When applied it assigns the [concludedLicense] to all matching findings.
```
Comment (mnonnenmacher) [inline model/src/main/kotlin/config/LicenseFindingCuration.kt]: Can you rephrase this? It's really hard to read. A proposal:

A curation for license findings.

File: model/src/main/kotlin/config/LicenseFindingCuration.kt [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+     * [detectedLicense].
+     */
+    val detectedLicense: String,
+    /**
+     * The concluded license as SPDX expression or [com.here.ort.spdx.SpdxLicense.NONE] for no license,
+     * see https://spdx.org/spdx-specification-21-web-version#h.jxpfx0ykyb60.
+     */
+    val concludedLicense: String,
+    /**
+     * A comment explaining this [LicenseFindingsCuration].
```
Comment (mnonnenmacher) [inline model/src/main/kotlin/config/LicenseFindingCuration.kt]: Typo: [LicenseFindingCuration]

File: model/src/main/kotlin/config/LicenseFindingCuration.kt [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    /**
+     * A matcher for the detected license of a license finding, matches if the detected license equals
+     * [detectedLicense].
+     */
+    val detectedLicense: String,
+    /**
+     * The concluded license as SPDX expression or [com.here.ort.spdx.SpdxLicense.NONE] for no license,
+     * see https://spdx.org/spdx-specification-21-web-version#h.jxpfx0ykyb60.
+     */
+    val concludedLicense: String,
```
Comment (mnonnenmacher) [inline model/src/main/kotlin/config/LicenseFindingCuration.kt]: Why not use `SpdxExpression?` as type for this? This would make sure that the string is a valid SPDX expression. We do the same in `PackageCurationData`.

File: model/src/test/kotlin/config/LicenseFindingCurationTest.kt [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+                detected_license: "MIT"
+                concluded_license: "Apache-2.0"
+                """.trimIndent()
+
+            val curation = yamlMapper
+                .readValue<LicenseFindingCuration>(yaml)
+                .let { yamlMapper.writeValueAsString(it) }
+                .let { yamlMapper.readValue<LicenseFindingCuration>(it) }
+
+            curation.startLines shouldBe listOf(1, 2, 3, 5)
```
Comment (mnonnenmacher) [inline model/src/test/kotlin/config/LicenseFindingCurationTest.kt]: Why not also verify that the other values are parsed correctly?

File: model/src/main/kotlin/util/FindingCurationMatcher.kt [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    fun applyAll(
+        findings: Collection<LicenseFinding>,
+        curations: Collection<LicenseFindingCuration>
+    ): List<LicenseFinding> =
+        findings.mapNotNull { finding ->
+            curations.find { matches(finding, it) }.let {
+                if (it != null) apply(finding, it)
+                else finding
+            }
+        }
```
Comment (mnonnenmacher) [inline model/src/main/kotlin/util/FindingCurationMatcher.kt]: The apply functions do not fit well into the matcher class. Why not put them in to the `LicenseFindingCuration` or the `Curations` class?

File: model/src/main/kotlin/config/LicenseFindingCuration.kt [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    val comment: String = ""
+) {
+    init {
+        require(startLines.all { it >= 0 }) {
+            "All values for start lines must not be negative."
+        }
+        require(lineCount == null || lineCount >= 0) {
+            "The value for line count must not be negative."
+        }
+        require(detectedLicense.isNotEmpty()) {
```
Comment (mnonnenmacher) [inline model/src/main/kotlin/config/LicenseFindingCuration.kt]: Maybe `isNotBlank()` as blank strings are also not valid? Same below.

File: model/src/main/kotlin/OrtResult.kt [hunk 8]
```diff
@@ -175,7 +177,11 @@ data class OrtResult(
         val project = getProject(id)
 
         scanResultsById[id].orEmpty().flatMap {
-            FindingsMatcher().match(it.summary.licenseFindings, it.summary.copyrightFindings)
+            FindingCurationMatcher()
+                .applyAll(it.summary.licenseFindings, getLicenseFindingsCurations(id))
+                .let { findings -> FindingsMatcher().match(findings, it.summary.copyrightFindings)
+                .toSortedSet()
+            }
```
Comment (mnonnenmacher) [inline model/src/main/kotlin/OrtResult.kt]: Formatting is broken here. Should be 

```kotlin
.let { findings ->
    FindingsMatcher().match(findings, it.summary.copyrightFindings).toSortedSet()
}
```
or single line.

File: model/src/main/kotlin/OrtResult.kt [hunk 9]
```diff
@@ -421,6 +427,10 @@ data class OrtResult(
     @JsonIgnore
     fun getExcludes(): Excludes = repository.config.excludes ?: Excludes()
 
+    private fun getLicenseFindingsCurations(id: Identifier): List<LicenseFindingCuration> =
+        if (projects.containsKey(id)) repository.config.curations?.licenseFindings.orEmpty()
+        else emptyList()
```
Comment (mnonnenmacher) [inline model/src/main/kotlin/OrtResult.kt]: Or use braces for if/else if it does not fit on a single line.

File: docs/Configuration.md [hunk 10]
```diff
@@ -162,6 +162,27 @@ resolutions:
 For details of the available reasons, see
 [RuleViolationResolutionReason.kt](../model/src/main/kotlin/config/RuleViolationResolutionReason.kt)
 
+### Curations
+
+#### License Findings
```
Comment (mnonnenmacher) [inline docs/Configuration.md]: Rather use `### Curating license findings` for consistency with the wording in other headers.

File: docs/Configuration.md [hunk 11]
```diff
@@ -162,6 +162,27 @@ resolutions:
...
+
+Scan result represent the detected licenses as collection of license findings. A single `LicenseFinding` is represented
```
Comment (mnonnenmacher) [inline docs/Configuration.md]: "**A** scan result represent**s** ... as **a** collection..."

File: docs/Configuration.md [hunk 12]
```diff
@@ -162,6 +162,27 @@ resolutions:
 For details of the available reasons, see
 [RuleViolationResolutionReason.kt](../model/src/main/kotlin/config/RuleViolationResolutionReason.kt)
 
+### Curations
+
+#### License Findings
+
+Scan result represent the detected licenses as collection of license findings. A single `LicenseFinding` is represented 
+as tuple: `(license-Id, file-path, start-line, end-line)`. A `LicenseFindingCuration` when applied changes the license-Id
```
Comment (mnonnenmacher) [inline docs/Configuration.md]: "Applying a `licenseFindingCuration` changes the..."

File: docs/Configuration.md [hunk 13]
```diff
@@ -162,6 +162,27 @@ resolutions:
...
+of any `LicenseFinding` license finding or eliminates the `LicenseFinding` in case the license is set to `NONE`.
```
Comment (mnonnenmacher) [inline docs/Configuration.md]: Duplication: "`LicenseFinding` license finding"

File: docs/Configuration.md [hunk 14]
```diff
@@ -162,6 +162,27 @@ resolutions:
...
+
+Changing similar license findings in all `.cpp` files within the `src` directory from to `Apache-2.0` could be done like
```
Comment (mnonnenmacher) [inline docs/Configuration.md]: "...from **`Apache-1.1`** to `Apache-2.0` **can**..."
Maybe rephrase: "As an example, the following curation would replace similar findings of `Apache-1.1` with `Apache-2.0` in all `.cpp` files in the `src` directory:"

File: docs/Configuration.md [hunk 15]
```diff
@@ -162,6 +162,27 @@ resolutions:
...
+```yaml
+  - path: "src/**/*.cpp"
+    start_lines: "3"
+    line_count: 11
+    detected_license: "Apache-1.1"
+    concluded_license: "Apache-1.0"
+ ```
+
+For details of the specification, see 
+[LicenseFindingCuration.kt](../model/src/main/kotlin/config/LicenseFindingCuration.kt)
```
Comment (mnonnenmacher) [inline docs/Configuration.md]: Dot at end of sentence missing.

File: model/src/main/kotlin/config/LicenseFindingCuration.kt (hunk 1 above)
Comment (fviernau) [reply model/src/main/kotlin/config/LicenseFindingCuration.kt]: Done.

File: model/src/main/kotlin/config/Curations.kt [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+ */
+
+package com.here.ort.model.config
+
+import com.fasterxml.jackson.annotation.JsonInclude
+
+/**
+ * Curations for artifacts in a repository.
+ */
+data class Curations(
```
[the remark this answers is not in the document]
Comment (fviernau) [reply model/src/main/kotlin/config/Curations.kt]: As discussed let's keep it as is now, as it is consistent with the rest of RepositoryConfiguration.

File: model/src/main/kotlin/config/Curations.kt [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+    /**
+     * Curations for license findings.
+     */
+    @JsonInclude(JsonInclude.Include.NON_EMPTY)
+    val licenseFindings: List<LicenseFindingCuration>
```
[the remark this answers is not in the document]
Comment (fviernau) [reply model/src/main/kotlin/config/Curations.kt]: As discussed let's keep it as is now, as it is consistent with the rest of RepositoryConfiguration.

File: model/src/main/kotlin/config/LicenseFindingCuration.kt (hunk 2 above)
Comment (fviernau) [reply model/src/main/kotlin/config/LicenseFindingCuration.kt]: Done.

File: docs/Configuration.md (hunk 11 above)
Comment (fviernau) [reply docs/Configuration.md]: Done.

File: docs/Configuration.md (hunk 12 above)
Comment (fviernau) [reply docs/Configuration.md]: Done.

File: docs/Configuration.md (hunk 13 above)
Comment (fviernau) [reply docs/Configuration.md]: Done.

File: docs/Configuration.md (hunk 14 above)
Comment (fviernau) [reply docs/Configuration.md]: Done.

File: docs/Configuration.md (hunk 15 above)
Comment (fviernau) [reply docs/Configuration.md]: Done.

File: model/src/main/kotlin/config/LicenseFindingCuration.kt (hunk 3 above)
Comment (fviernau) [reply model/src/main/kotlin/config/LicenseFindingCuration.kt]: Done.

File: model/src/test/kotlin/config/LicenseFindingCurationTest.kt (hunk 5 above)
Comment (fviernau) [reply model/src/test/kotlin/config/LicenseFindingCurationTest.kt]: Because the test is intentionally only covering ```deserializing``` the ```start_lines``` as we use a custom ```com.fasterxml.jackson.databind.util.Converter``` for. All other properties are serialized without any ```custom``` code an thus is supposed to be covered by tests in the ```jackson``` project.

File: docs/Configuration.md (hunk 10 above)
Comment (fviernau) [reply docs/Configuration.md]: Done.

File: model/src/main/kotlin/OrtResult.kt (hunk 9 above)
Comment (fviernau) [reply model/src/main/kotlin/OrtResult.kt]: Done.

File: model/src/main/kotlin/OrtResult.kt (hunk 8 above)
Comment (fviernau) [reply model/src/main/kotlin/OrtResult.kt]: Done.

File: model/src/main/kotlin/config/LicenseFindingCuration.kt (hunk 7 above)
Comment (fviernau) [reply model/src/main/kotlin/config/LicenseFindingCuration.kt]: Done.

File: model/src/main/kotlin/util/FindingCurationMatcher.kt (hunk 6 above)
Comment (fviernau) [reply model/src/main/kotlin/util/FindingCurationMatcher.kt]: I think it does fit quite well as the purpose of this class is ```applying curations```. Would the issue you mention be fixed by renaming the class ? I don't have a nice idea though, how about ```LicenseFindingCurator```?

I didn't put it into the ```Curation``` class because:
1. The responsibility of applying curations seemed like complex enough to put it into a separate class, e.g. it seemed to be one reposibility which can be considered in isolation. 
2. Coupling
  - no dependencies between ```LicenseFindingCuration```and ```LicenseFinding```
  - no dependencies from ```LicenseFindingCuration``` to the matching logic

File: model/src/main/kotlin/config/LicenseFindingCuration.kt (hunk 4 above)
Comment (fviernau) [reply model/src/main/kotlin/config/LicenseFindingCuration.kt]: note from our discussion: ```detected``` licenses are still handled as ```String``` not as ```SPDX-Expression```.  As the applied curations are handled as ```detected license``` what you are asking for would require to handle ```detected licenses``` as ```SPDX Expression``` which is however a bigger topic which makes sense to be handled separatly.

File: model/src/test/kotlin/config/LicenseFindingCurationTest.kt (hunk 5 above)
Comment (mnonnenmacher) [reply model/src/test/kotlin/config/LicenseFindingCurationTest.kt]: There are many ways for us to mess with the (de)serialization, e.g. by changing the global Jackson configuration, so I believe there's nothing wrong with testing a few more values.

File: model/src/main/kotlin/util/FindingCurationMatcher.kt (hunk 6 above)
Comment (mnonnenmacher) [reply model/src/main/kotlin/util/FindingCurationMatcher.kt]: Ok, let's merge it like that, I can propose a refactoring if I feel it's required.
